stdlib/src/builtin/string.mojo

```diff
@@ -388,6 +388,134 @@ fn atol(str: String, base: Int = 10) raises -> Int:
+
+    For example, `atof("2.25")` returns `2.25`. If the given string cannot be parsed
+    as an float value, an error is raised. For example, `atof("hi")` raises an
+    error.
+
+    Args:
+        str_ref: A string to be parsed as a floating point.
+
+    Returns:
+        An float value that represents the string, or otherwise raises.
+    """
+    if not str_ref:
+        raise Error(_atof_error(str_ref))
+
+    var result: Float64 = 0.0
+    var exponent: Int = 0
+    var is_negative: Bool = False
+
+    alias ord_0 = ord("0")
+    alias ord_9 = ord("9")
+    alias ord_dot = ord(".")
+    alias ord_plus = ord("+")
+    alias ord_minus = ord("-")
+    alias ord_f = ord("f")
+    alias ord_F = ord("F")
+    alias ord_e = ord("e")
+    alias ord_E = ord("E")
+
+    var start: Int = 0
+    var str_len = len(str_ref)
+    var buff = str_ref.unsafe_ptr()
+
+    var c: Int = 0
+    # skip leading spaces and read sign
+    for pos in range(start, str_len):
+        c = int(buff[pos])
+        if isspace(c):
+            pass
+        elif c == ord_minus:
+            if is_negative:
+                raise Error(_atof_error(str_ref))
+            is_negative = True
+        else:
+            break
+        start += 1
+    # read before dot
+    for pos in range(start, str_len):
+        c = int(buff[pos])
+        if ord_0 <= c <= ord_9:
+            result = result * 10.0 + (c - ord_0)
+            start += 1
+        else:
+            break
+    # if dot -> read after dot
+    if c == ord_dot:
+        start += 1
+        for pos in range(start, str_len):
+            c = int(buff[pos])
+            if ord_0 <= c <= ord_9:
+                result = result * 10.0 + (c - ord_0)
+                exponent -= 1
+            else:
+                break
+            start += 1
+    # if e/E -> read scientific notation
+    if c == ord_e or c == ord_E:
+        start += 1
+        var sign: Int = 1
+        var shift: Int = 0
+        var has_number: Bool = False
+        for pos in range(start, str_len):
+            c = int(buff[pos])
+            if c == ord_plus:
+                pass
+            elif c == ord_minus:
+                sign = -1
+            elif ord_0 <= c <= ord_9:
+                has_number = True
+                shift = shift * 10 + (c - ord_0)
+            else:
+                break
+            start += 1
+        exponent += sign * shift
+        if not has_number:
+            raise Error(_atof_error(str_ref))
+    # check if the string is fully parsed
+    if start != str_len:
+        if c == ord_f or c == ord_F:  # f/F at end is allowed
+            start += 1
+        for pos in range(start, str_len):
+            c = int(buff[pos])
+            if not isspace(c):
+                raise Error(_atof_error(str_ref))
+    # apply shift
+    var shift: Int = 10 ** abs(exponent)
+    if exponent > 0:
+        result *= shift
+    if exponent < 0:
+        result /= shift
```

Comment:
Since `result` is a `Float64` this will implicitly convert `shift` anyway. Does this work?
```suggestion
    var result *= 10.0 ** exponent
```

Reply:
the problem here is that mojo rounds the floats differently, and doing a division for negative exponents is closer to how mojo does it.
e.g., `-0.3`
mojo stores it as `-0.29999999999999999`
while the multiply method (3.0 * 0.1) returns `-0.30000000000000004`
the division method (3.0 / 10.0) produces `-0.29999999999999999`

so the result is different depending on doing a `* 10` or `/ 0.1` or vice versa

A similar strange situation is the following
```python
print(233.0 / 1000.0) #prints 0.23299999999999998
var x: Float64 = 233.0
var div: Float64 = 1000.0
print(x/div) #prints 0.23300000000000001
```
If this is a compiler issue and we can fix that then this fix would be definitely the better option

Reply:
Okay, this makes sense actually. Can you please ensure we have at least one test case that hits this? Please leave a comment both here and at that test case to explain this. Thank you!